### PR TITLE
Configuration JSON schema

### PR DESCRIPTION
**Pull request by laudibert**
Feature asked in #73.

Not sure if the schema should have `"additionalProperties": false` to help users realize when a field name was mistyped or if it's too constraining for when the schema might not be up to date with the code.

### CODE REVIEW

**Comment by laudibert** [comment]
Also, I automatically filled the `"description"` of each field with whatever was on the `README.md` but I realized only afterwards that it might be a pain to maintain these pieces of documentation in both the `README.md` and the schema. Ideally though, the schema should have them though I guess.

File: src/components/app.context.ts [hunk 1]
```diff
@@ -16,10 +15,7 @@ export const AppContext = React.createContext<IAppContext>({
   config: null,
   activePage: null,
   error: null,
-  setError: (error: string | null) => {
```
Comment (dsternlicht) [inline src/components/app.context.ts]: 😄

File: README.md [hunk 2]
```diff
@@ -412,6 +412,7 @@ Here's a list of available display field types:
 * `url` - will render an anchor element with a clickable link 
 * `image` - will render an image from the url 
 * `colorbox` - will render a #RRGGBB hex string as an 80 x 20 pixel colored rectangle, overlaid with the hex color string
+* `boolean` - will green or red dot
```
Comment (dsternlicht) [inline README.md:415]: * will **be**

**Comment by dsternlicht** [comment]
> Feature asked in #73.
> 
> Not sure if the schema should have `"additionalProperties": false` to help users realize when a field name was mistyped or if it's too constraining for when the schema might not be up to date with the code.

I think it's definitely should be part of the schema.

**Comment by dsternlicht** [comment]
> Also, I automatically filled the `"description"` of each field with whatever was on the `README.md` but I realized only afterwards that it might be a pain to maintain these pieces of documentation in both the `README.md` and the schema. Ideally though, the schema should have them though I guess.

You did right. It's a bit pain, but very important.

**Comment by dsternlicht** [comment]
@laudibert That's an awesome PR for an extremely important feature! Kudos!

File: src/components/app.context.ts (hunk 1 above)
Comment (laudibert) [reply src/components/app.context.ts]: 😅

**Comment by laudibert** [comment]
@dsternlicht, it should be ok now. The only problem that I encountered without any solution is that the `"additionalProperties": false` does not work with the keyword `"allOf"`that I used to manage inheritance for methods schema definitions. Therefore it is not set to `false` for the methods.

**Comment by dsternlicht** [comment]
I'm sorry to ask, but what's the meaning of `"additionalProperties": false`? It's not to allow other properties than the one you defined in the schema?

What's the `"allOff"` keyword?

**Comment by laudibert** [comment]
> I'm sorry to ask, but what's the meaning of `"additionalProperties": false`? It's not to allow other properties than the one you defined in the schema?

Yes but it works only one level at a time. Therefore it has to be defines for any object throughout the configuration. Not just at the root.

> What's the `"allOff"` keyword?

It allows to combine schemas for one schema definition. See [here](https://json-schema.org/understanding-json-schema/reference/combining.html#allof) for a good explanation. In the RESTool case, it allowed me to define the common properties of all methods in one place and then combine it with the specific of each method.

**Comment by dsternlicht** [comment]
Got. Does schemas support inheritance from another file as well?

**Comment by laudibert** [comment]
Yes, definitions can come from other files all loaded at the start of the app. I did not choose to do this though as I thought it might not be practical for people that are asking for the full schema.

I could change it though.

**Comment by dsternlicht** [comment]
@laudibert no need, just ask for my own general knowledge :)

**Pull request merged**